File: src/main/java/org/ahmadsoft/ropes/impl/RepeatedCharacterSequence.java
/*
 * Copyright (C) 2024
 *  - Originally Copyright (C) 2007 Amin Ahmad.
 * Licenced under GPL
 */
package org.ahmadsoft.ropes.impl;

import java.util.Arrays;

/**
 * A character sequence defined by a character
 * and a repeat count.
 * @author Amin Ahmad
 */
public class RepeatedCharacterSequence implements CharSequence {
	
	private char character;
	private int repeat;

	public RepeatedCharacterSequence(char character, int repeat) {
		super();
		this.character = character;
		this.repeat = repeat;
	}

	@Override
	public char charAt(int index) {
		return getCharacter();
	}

	@Override
	public int length() {
		return repeat;
	}

	@Override
	public CharSequence subSequence(int start, int end) {
		return new RepeatedCharacterSequence(getCharacter(), end - start);
	}
	
	@Override
	public String toString() {
		char[] result = new char[repeat];
		Arrays.fill(result, character);
		return new String(result);
	}

	/**
	 * Returns the character used to construct this sequence.
	 * @return the character used to construct this sequence.
	 */
	public char getCharacter() {
		return character;
	}

}

File: src/main/kotlin/net/ropes4k/impl/RepeatedCharacterSequence.kt
/*
 * Copyright (C) 2024 James Richardson
 *  - Originally Copyright (C) 2007 Amin Ahmad.
 * Licenced under GPL
 */
package net.ropes4k.impl

import java.util.Arrays

/**
 * A character sequence defined by a character
 * and a repeat count.
 */
internal class RepeatedCharacterSequence(
    private val character: Char,
    private val repeat: Int
) : CharSequence {
    override fun get(index: Int): Char {
        return character
    }

    override val length = repeat


    override fun subSequence(startIndex: Int, endIndex: Int): CharSequence {
        return RepeatedCharacterSequence(character, endIndex - startIndex)
    }

    override fun toString(): String {
        val result = CharArray(repeat)
        Arrays.fill(result, character)
        return String(result)
    }
}
